### ISSUE TITLE
hive.write fails to drop existing partitions

### ISSUE THREAD

**Issue by dnguyendena** [DeNA/mobilize-server#49]
Hi Cassio,

I have Mobilize working great in JP cluster.
However, we recently found a problem in Mobilize JP : the hive.write command could not drop existing partitions in a table before overwriting the partitions.
I have tracked the execution process and get the following:
1. The main Hive query (i.e., the query specified in Runner) was successfully executed
2. I saw this temporary Hive script in the log (in /home/user/mobilize/ ) : 

select distinct partition_col from tmp_table;

and I believe that this script was executed (the directory containing this log was removed by Mobilize)
3. I could not see the final step in which Mobilize drop the partition of the target table and write the result

Here is the error : 
response
undefined method `downcase' for nil:NilClass
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/mobilize-hive-1.371/lib/mobilize-hive/handlers/hive.rb:341:in`block (2 levels) in hql_to_table'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/mobilize-hive-1.371/lib/mobilize-hive/handlers/hive.rb:340:in `each'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/mobilize-hive-1.371/lib/mobilize-hive/handlers/hive.rb:340:in`map'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/mobilize-hive-1.371/lib/mobilize-hive/handlers/hive.rb:340:in `block in hql_to_table'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/mobilize-hive-1.371/lib/mobilize-hive/handlers/hive.rb:339:in`map'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/mobilize-hive-1.371/lib/mobilize-hive/handlers/hive.rb:339:in `hql_to_table'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/mobilize-hive-1.371/lib/mobilize-hive/handlers/hive.rb:466:in`write_by_stage_path'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/mobilize-base-1.371/lib/mobilize-base/models/stage.rb:59:in `perform'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/resque-1.24.0/lib/resque/job.rb:125:in`perform'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/resque-1.24.0/lib/resque/worker.rb:193:in `perform'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/resque-1.24.0/lib/resque/worker.rb:144:in`block (2 levels) in work'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/resque-1.24.0/lib/resque/worker.rb:264:in `fork'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/resque-1.24.0/lib/resque/worker.rb:264:in`fork'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/resque-1.24.0/lib/resque/worker.rb:141:in `block in work'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/resque-1.24.0/lib/resque/worker.rb:132:in`loop'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/resque-1.24.0/lib/resque/worker.rb:132:in `work'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/mobilize-base-1.371/lib/mobilize-base/tasks.rb:21:in`block (2 levels) in <top (required)>'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/rake-10.1.0/lib/rake/task.rb:236:in `call'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/rake-10.1.0/lib/rake/task.rb:236:in`block in execute'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/rake-10.1.0/lib/rake/task.rb:231:in `each'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/rake-10.1.0/lib/rake/task.rb:231:in`execute'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/rake-10.1.0/lib/rake/task.rb:175:in `block in invoke_with_call_chain'
/opt/analytics/ruby/lib/ruby/1.9.1/monitor.rb:211:in`mon_synchronize'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/rake-10.1.0/lib/rake/task.rb:168:in `invoke_with_call_chain'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/rake-10.1.0/lib/rake/task.rb:161:in`invoke'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/rake-10.1.0/lib/rake/application.rb:149:in `invoke_task'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/rake-10.1.0/lib/rake/application.rb:106:in`block (2 levels) in top_level'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/rake-10.1.0/lib/rake/application.rb:106:in `each'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/rake-10.1.0/lib/rake/application.rb:106:in`block in top_level'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/rake-10.1.0/lib/rake/application.rb:115:in `run_with_threads'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/rake-10.1.0/lib/rake/application.rb:100:in`top_level'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/rake-10.1.0/lib/rake/application.rb:78:in `block in run'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/rake-10.1.0/lib/rake/application.rb:165:in`standard_exception_handling'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/rake-10.1.0/lib/rake/application.rb:75:in `run'
/opt/analytics/ruby/lib/ruby/gems/1.9.1/gems/rake-10.1.0/bin/rake:33:in`<top (required)>'
/opt/analytics/ruby/bin/rake:23:in `load'
/opt/analytics/ruby/bin/rake:23:in`<main>'

When I tracking the same query in US cluster, I found that it works well so I believe that I missed some parameters in Mobilize hive configuration

Here is my config file :  
production:
  prepends:
    - "hive.stats.autogather=false"
  clusters:
    prod_cluster:
      max_slots: 5
      output_db: mobilize
      exec_path: /usr/bin/hive

Could you please help me debug this problem?

Thanks,
Duc.

**Comment by c4ssio** [comment]
Hi Duc,

From https://github.com/DeNA/mobilize-server/wiki/Jobs:-Handlers-and-Methods, hive.write section:

To write the Source data to the Target, Mobilize will: 
1) write the results from the source to a temporary table 
2) calculate the different possible permutations 
3) delete those partitions from the target table, if any 
4) write the temporary table to the target table

Looks like the issue is in step 2, which is not figuring out which partitions exist and therefore cannot delete them.

I don't see any obvious issues, so we should step through the code together to figure out what the problem is.  Find me on gtalk at mobilize.server@gmail.com and I will help you.

Thanks,

Cassio

**Comment by dnguyendena** [comment]
Cassio helped me to debug the problem.
We found that the root cause was in line 341 of this file : 

https://github.com/DeNA/mobilize-hive/blob/7da836ccd784f172e077caa483e13622c5a3234d/lib/mobilize-hive/handlers/hive.rb

If the partition name contains only one character (e.g., 'd'), then seems that the code trims it to empty. That causes existing partitions in the target table not correctly dropped.
I will send Ryosuke a bug report on this.

Thank you Cassio for your great help!.

Duc.

**Comment by dnguyendena** [comment]
Here is the reason for this trimming : 
In US cluster (Hive 0.7.0), the column name in query result contains back-quotes :  

``` sql
set hive.cli.print.header=true;
select `d` from some_table where d="2013-08-15" limit 1; (1)
select d from some_table where d="2013-08-15" limit 1;   (2)
```

The above two queries produce different results : 

```
Result of (1) is 
`d`
2013-08-15

Result of (2) is
d
2013-08-15
```

In JP cluster, we use Hive 0.10.0 and seems that this bug in Hive has been fixed, the above two queries produce identical results :  

```
(1)
d
2013-08-15
(2)
d
2013-08-15
```

Many thanks to Ryosuke for finding out the reason and for the quick fix!

Best,
Duc.

**Comment by riywo** [comment]
Thank you for the reporting. I released v1.376.

https://rubygems.org/gems/mobilize-hive/versions/1.376
